### PR TITLE
dbqn: init at unstable-2021-09-03

### PR DESCRIPTION
**Pull request by sternenseemann**
###### Motivation for this change

Add dzaima/BQN, a BQN implementation. This may especially be interesting as another way to bootstrap CBQN (see PR #135665).

Doesn't include the processing based application at the moment, only the (CLI) language implementation itself.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/interpreters/dbqn/default.nix [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+, dbqn
+}:
+
+let
+  java = if useNativeImage then graalvm11-ce else jre;
+  useStdenv = if useNativeImage then stdenv else stdenvNoCC;
+in
+
+useStdenv.mkDerivation {
+  pname = "dBQN";
```
Comment (AndersonTorres) [inline pkgs/development/interpreters/dbqn/default.nix:22]: Write this in all-lowercase.

File: pkgs/development/interpreters/dbqn/default.nix [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    # Only prints results, so we need to check success
+    "$out/bin/dbqn" ./test/test | tee /dev/stderr | grep -v "Passed all" && exit 1
+
+    runHook postInstallCheck
+  '';
+
+  meta = {
+    description = "A BQN implementation based on dzaima/APL"
+      + lib.optionalString useNativeImage ", natively compiled as a GraalVM native image";
```
Comment (AndersonTorres) [inline pkgs/development/interpreters/dbqn/default.nix:85]: `natively ` is redundant.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/development/interpreters/dbqn/default.nix [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    install -Dm644 BQN.jar -t "$out/share/dbqn"
+
+    # script interpreter
+    makeWrapper "${lib.getBin java}/bin/java" "$out/bin/dbqn" \
+      --add-flags "-jar $out/share/dbqn/BQN.jar -f"
+
+    # repl
+    makeWrapper "${lib.getBin java}/bin/java" "$out/bin/dbqn-repl" \
+      --add-flags "-jar $out/share/dbqn/BQN.jar -r"
+  '' + lib.optionalString useNativeImage ''
```
[the remark this answers is not in the document]
Comment (sternenseemann) [reply pkgs/development/interpreters/dbqn/default.nix]: Done.

File: pkgs/development/interpreters/dbqn/default.nix [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  installCheckPhase = ''
+    runHook preInstallCheck
+
+    # Only prints results, so we need to check success
+    "$out/bin/dbqn" ./test/test | tee /dev/stderr | grep -v "Passed all" && exit 1
+
+    runHook postInstallCheck
+  '';
+
+  meta = {
```
[the remark this answers is not in the document]
Comment (sternenseemann) [reply pkgs/development/interpreters/dbqn/default.nix]: Doesn't make a difference.

File: pkgs/development/interpreters/dbqn/default.nix (hunk 1 above)
Comment (sternenseemann) [reply pkgs/development/interpreters/dbqn/default.nix]: Done.

File: pkgs/development/interpreters/dbqn/default.nix (hunk 2 above)
Comment (sternenseemann) [reply pkgs/development/interpreters/dbqn/default.nix]: Thanks, removed.

**Pull request closed without merging**